mlbench_core/optim/pytorch/optim.py

```diff
@@ -499,12 +499,302 @@ def step(self, closure=None):
+
+        # update parameters
+        i = 0
+        for group in self.param_groups:
+            weight_decay = group["weight_decay"]
+            momentum = group["momentum"]
+            dampening = group["dampening"]
+            nesterov = group["nesterov"]
+
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                d_p = grads[i]
+                i += 1
+                if weight_decay != 0:
+                    d_p.add_(weight_decay, p.data)
+                if momentum != 0:
+                    param_state = self.state[p]
+                    if "momentum_buffer" not in param_state:
+                        buf = param_state["momentum_buffer"] = torch.clone(d_p).detach()
+                    else:
+                        buf = param_state["momentum_buffer"]
+                        buf.mul_(momentum).add_(1 - dampening, d_p)
+                    if nesterov:
+                        d_p = d_p.add(momentum, buf)
+                    else:
+                        d_p = buf
+
+                p.data.add_(-group["lr"], d_p)
+
+        return loss
+
+    def reduce(self, grad_in, grad_out, memory_out):
+        # Split the tensors into rank1-ones that will be reduced un-compressed
+        # and rank > 1 tensors that are compressed
+        rank1_tensors = [
+            (tensor, out, mem)
+            for tensor, out, mem in zip(grad_in, grad_out, memory_out)
+            if tensor.ndimension() <= 1
+        ]
+        high_rank_tensors = [
+            (tensor, out, mem)
+            for tensor, out, mem in zip(grad_in, grad_out, memory_out)
+            if tensor.ndimension() > 1
+        ]
+
+        # We are building a rank-1 approximation of every tensor
+        # that can be interpreted as a matrix. Let the approximation be
+        # M = p q^T
+        # We are allocating consequtive memory for the p's and q's
+
+        memory_is_uninitialized = self.p_memory is None
+
+        p_total_size = 0
+        q_total_size = 0
+        for tensor, _, _ in high_rank_tensors:
+            matrix = tensor.view(tensor.shape[0], -1)
+            n, m = matrix.shape
+            rank = min(n, m, self.rank)
+            p_total_size += n * rank
+            q_total_size += m * rank
+        if self.p_memory is None:
+            self.p_memory = torch.empty(p_total_size, device=self.device)
+            self.q_memory = torch.empty(q_total_size, device=self.device)
+
+        # Find them again and make lists of pointers
+        ps = []
+        qs = []
+        p_idx = 0
+        q_idx = 0
+        for tensor, _, _ in high_rank_tensors:
+            matrix = tensor.view(tensor.shape[0], -1)
+            n, m = matrix.shape
+            rank = min(n, m, self.rank)
+            ps.append(self.p_memory[p_idx : p_idx + n * rank].view(n, rank))
+            qs.append(self.q_memory[q_idx : q_idx + m * rank].view(m, rank))
+            p_idx += n * rank
```

Comment:
Maybe it would be nice to abstract away this bookkeeping into a `pack` and `unpack` function (that operates without copying), if there is not already something like that in the codebase.